### PR TITLE
[CMAKE] Unify configs for architectures and make ARCH required

### PR DESCRIPTION
**Pull request by Extravert-ir**
config*.cmake files are constantly getting out of sync between
each other. Besides that, the parameters are not really dependent
on a target architecture, but rather on a compiler (except *ARCH)

This approach seems to be more future-prone, and allows to see
all the options in one file (there are really not that much of them)

### CODE REVIEW

File: sdk/cmake/config.cmake [hunk 1]
```diff
@@ -80,13 +104,13 @@ set(GENERATE_DEPENDENCY_GRAPH FALSE CACHE BOOL
...
-    # RTC are incompatible with compiler optimizations.
-    cmake_dependent_option(RUNTIME_CHECKS "Whether to enable runtime checks on MSVC" ON
-                           "CMAKE_BUILD_TYPE STREQUAL Debug" OFF)
+    set(_PREFAST_ FALSE CACHE BOOL
+    "Whether to enable PREFAST while compiling.")
+    set(_VS_ANALYZE_ FALSE CACHE BOOL
+    "Whether to enable static analysis while compiling.")
+        # RTC are incompatible with compiler optimizations.
+        cmake_dependent_option(RUNTIME_CHECKS "Whether to enable runtime checks on MSVC" ON
+                               "CMAKE_BUILD_TYPE STREQUAL Debug" OFF)
```
Comment (binarymaster) [inline sdk/cmake/config.cmake:113]: Indentation :)
```suggestion
    # RTC are incompatible with compiler optimizations.
    cmake_dependent_option(RUNTIME_CHECKS "Whether to enable runtime checks on MSVC" ON
                           "CMAKE_BUILD_TYPE STREQUAL Debug" OFF)
```

File: sdk/cmake/gcc.cmake [hunk 2]
```diff
@@ -90,11 +90,7 @@ if(NOT CMAKE_BUILD_TYPE STREQUAL "Release")
 endif()
 
 # Tuning
-if(ARCH STREQUAL "i386")
-    add_compile_options(-march=${OARCH} -mtune=${TUNE})
-else()
-    add_compile_options(-march=${OARCH})
-endif()
+add_compile_options(-march=${OARCH} -mtune=${TUNE})
```
Comment (learn-more) [inline sdk/cmake/gcc.cmake:93]: Is it going to cry about the empty `-mtune=`?

Comment (Extravert-ir) [reply sdk/cmake/gcc.cmake:93]: I've defined `TUNE` for all platforms

File: sdk/cmake/config.cmake [hunk 3]
```diff
@@ -1,14 +1,38 @@
...
+    set(SARCH "pc" CACHE STRING
+    "Sub-architecture to build for. Specify one of:
+     pc pc98 xbox")
+elseif(ARCH STREQUAL "amd64")
+    set(SARCH "" CACHE STRING
+    "Sub-architecture to build for.")
+elseif(ARCH STREQUAL "arm")
+    set(SARCH "omap3-zoom2" CACHE STRING
+    "Sub-architecture (board) to build for. Specify one of:
+     kurobox versatile omap3-zoom2 omap3-beagle")
```
Comment (SergeGautherie) [inline sdk/cmake/config.cmake:12]: Add
```cmake
else()
    message(FATAL_ERROR "Unsupported ARCH: ${ARCH}")
```
To `OARCH` and `TUNE` too.

File: toolchain-gcc.cmake [hunk 4]
```diff
@@ -1,13 +1,6 @@
 
-if(NOT ARCH)
-    set(ARCH i386)
-endif()
-
-# Default to Debug for the build type
```
Comment (SergeGautherie) [inline toolchain-gcc.cmake:6]: `gcc` and `clang`: Now 'empty' instead of `Debug` :-/

File: toolchain-msvc.cmake [hunk 5]
```diff
@@ -1,13 +1,6 @@
 
-if(NOT ARCH)
-    set(ARCH i386)
-endif()
-
-# Default to Debug for the build type
```
Comment (SergeGautherie) [inline toolchain-msvc.cmake:6]: `msvc`: Still defaults to `Debug`, somehow :-)
`msbuild`: Now 'empty' instead of `Debug` :-/

File: toolchain-msvc.cmake [hunk 6]
```diff
@@ -1,13 +1,6 @@
...
-if(NOT DEFINED CMAKE_BUILD_TYPE)
-    set(CMAKE_BUILD_TYPE "Debug" CACHE STRING
-        "Choose the type of build, options are: None(CMAKE_CXX_FLAGS or CMAKE_C_FLAGS used) Debug Release RelWithDebInfo MinSizeRel.")
-endif()
+# pass variables necessary for the toolchain (needed for try_compile)
+set(CMAKE_TRY_COMPILE_PLATFORM_VARIABLES CMAKE_BUILD_TYPE ARCH USE_CLANG_CL)
 
 # the name of the target operating system
 set(CMAKE_SYSTEM_NAME Windows)
```
Comment (SergeGautherie) [inline toolchain-msvc.cmake:6]: `CMAKE_SYSTEM_NAME` seems unrelated to compilers: to be merged elsewhere?

File: CMakeLists.txt [hunk 7]
```diff
@@ -29,7 +29,7 @@ set(CMAKE_CXX_STANDARD 11)
 #set_property(GLOBAL PROPERTY RULE_MESSAGES OFF)
 
 if(NOT ARCH)
-    set(ARCH i386)
+    message(FATAL_ERROR "Target architecture (ARCH) is not defined. Please, choose one of: i386, amd64, arm")
```
[the remark this answers is not in the document]
Comment (Extravert-ir) [reply CMakeLists.txt:33]: I'm sure the existence check is enough

File: toolchain-gcc.cmake (hunk 4 above)
Comment (Extravert-ir) [reply toolchain-gcc.cmake:6]: It looks like we need to eliminate all usages of `CMAKE_BUILD_TYPE` in order to avoid bugs...
https://stackoverflow.com/questions/66079007/having-conditional-statements-on-build-type-variable-a-good-design
